### PR TITLE
[RFC] drivers: mipi_dbi: add support for mipi_dbi_configure_te

### PR DESCRIPTION
**Pull request by danieldegrasse**
Many MIPI DBI displays support a "tearing effect" signal, which can be
configured to signal each v-sync or h-sync interval. This signal can be
used by the MIPI DBI controller to synchronize writes with the
controller, and avoid tearing effects on the screen (which occur when
the write pointer from the MCU overlaps with the panel's read pointer in
the display controller's graphics RAM).

Add the `mipi_dbi_configure_te` API, which allows display controllers to
configure MIPI DBI controller to wait for a TE edge before streaming
display data. Allow the tearing enable parameters to be configured via
devicetree settings, since these will vary based on the MIPI DBI
controller and display controller in use.

For an example of how this signal can be used, see the write diagrams within the ST7796S datasheet (section 10.8.3):

image

image

~Depends on #81245 for some fixes to the LCDIC driver~

Todo:

- [x] Validate changes on `frdm_mcxn947` with `lcd_par_s035_8080` shield
- [x] Validate changes on `rd_rw612_bga` with `adafruit_2_8_tft_touch_v2` shield

### CODE REVIEW

File: dts/bindings/mipi-dbi/mipi-dbi-device.yaml [hunk 1]
```diff
@@ -34,3 +34,26 @@ properties:
...
+      MIPI DBI tearing enable signal mode. Use the macros, not the actual enum
+      value. Defaults to disabled.
+      Here is the concordance list (see dt-bindings/mipi_dbi/mipi_dbi.h)
+        0     MIPI_DBI_TE_NO_EDGE
+        1     MIPI_DBI_TE_RISING_EDGE
+        2     MIPI_DBI_TE_FALLING_EDGE
+    enum:
+      - 0
+      - 1
+      - 2
```
Comment (pdgendt) [inline dts/bindings/mipi-dbi/mipi-dbi-device.yaml:51]: Use a string property together with `DT_INST_STRING_UPPER_TOKEN` to make it more explicit and clear.

Comment (danieldegrasse) [reply dts/bindings/mipi-dbi/mipi-dbi-device.yaml]: Thanks for this idea- updated in latest push

File: include/zephyr/drivers/mipi_dbi.h [hunk 2]
```diff
@@ -293,6 +320,45 @@ static inline int mipi_dbi_release(const struct device *dev,
...
+ * provided with the prior call, as this indicates the buffer being written
+ * in this call is the first buffer of a new frame.
+ *
+ * Note that most display controllers will need to enable the TE signal
+ * using vendor specific commands before the MIPI DBI controller can react
+ * to it.
+ *
+ * @param dev mipi dbi controller
+ * @param edge which edge of the TE signal to start transmitting on
+ * @param delay_us how many microseconds after TE edge to start transmission
```
Comment (erwango) [inline include/zephyr/drivers/mipi_dbi.h:344]: Some controllers also support TE source (DSI link or External pin). How would you propose to configure this ?

Comment (danieldegrasse) [reply include/zephyr/drivers/mipi_dbi.h:344]: Do you have an example I could look at? Not sure I follow here. Do you mean that the MIPI controller supports selecting the TE source between an on chip DSI peripheral and external pin?

Comment (erwango) [reply include/zephyr/drivers/mipi_dbi.h:344]: Yes, supported in https://www.st.com/resource/en/reference_manual/rm0399-stm32h745755-and-stm32h747757-advanced-armbased-32bit-mcus-stmicroelectronics.pdf.
See 34.16.1
Screenshot from 2024-11-26 16-25-28

Comment (danieldegrasse) [reply include/zephyr/drivers/mipi_dbi.h:344]: Thanks for pointing this out- one thing to note, this is a _DSI_ controller, not a _DBI_ controller- so it wouldn't fit within the MIPI DBI API itself.

Looking at this more, I do think we'd need something like this within the MIPI DSI API though- it looks like DSI controllers can either signal TE via a dedicated GPIO pin, or via a D-PHY trigger event on the DSI bus itself.

**Comment by danieldegrasse** [comment]
@Finomnis or @faxe1008, would either of you be willing to provide feedback on this PR?

**Comment by Finomnis** [comment]
Sure, but probably not within the next couple of days, due to short deadlines. Remind me again in a couple of days if I forget.

File: drivers/mipi_dbi/mipi_dbi_nxp_lcdic.c [hunk 3]
```diff
@@ -413,6 +426,24 @@ static int mipi_dbi_lcdic_write_display(const struct device *dev,
...
+			break;
+		default:
+			te_sync = LCDIC_TE_NO_SYNC;
+			break;
+		}
+		dev_data->new_frame = false;
+	} else if (!desc->frame_incomplete) {
+		/* Next frame will be a new one */
+		dev_data->new_frame = true;
+	}
```
Comment (Finomnis) [inline drivers/mipi_dbi/mipi_dbi_nxp_lcdic.c:445]: I'm a little confused here; the way this `if` statement is written, if every single frame is `frame_incomplete = false`, then the `te_sync` will only be active every other `write` call. Is that intended?

I feel like the `else` should be removed, and the `if (!desc->frame_incomplete)` should be executed always.

Feel free to correct my thoughts if I'm mistaken.

Comment (danieldegrasse) [reply drivers/mipi_dbi/mipi_dbi_nxp_lcdic.c:445]: No, I think you are right here. We always want the next call after one with `frame_incomplete = false` to trigger a TE wait. I'll update this.

Comment (danieldegrasse) [reply drivers/mipi_dbi/mipi_dbi_nxp_lcdic.c:447]: Updated

**Review by jfischer-no — APPROVED** [review]

**Comment by danieldegrasse** [comment]
Can't approve my own PR, reassigning to @dleach02 as this touches NXP drivers

**Review by rruuaanng — APPROVED** [review]

**Comment by Finomnis** [comment]
@danieldegrasse did you confirm with an osci that this is working properly?

**Comment by danieldegrasse** [comment]
> @danieldegrasse did you confirm with an osci that this is working properly?

I have not with latest push (as I no longer have access to the hardware), but yes I have verified the TE signal delays as expected with a signal analyzer using the LVGL sample. That didn't expose the bug you caught because LVGL will only set `frame_incomplete` to false once every 10 frames in the default configuration (where the framebuffer is 10% of the screen size)

**Review by dleach02 — APPROVED** [review]

**Pull request merged**